### ISSUE TITLE
O2.3.11 Add land/sea mask to ClimaCore domain

### ISSUE THREAD

**Issue by kmdeck** [CliMA/ClimaLand.jl#488]
## Purpose
For global runs, we currently set up a spherical shell domain (extruded FD) and create fields that are defined everywhere on the domain. The issues arise in three areas:
- Interpolation: land parameters are defined over land only. When we interpolate these to the model grid, we need to be aware of the mask so as not to get unphysical land parameters over the ocean. For example, the van genuchten n parameter must be > 1. If over the ocean, this is zero, a cell with a large fraction of ocean will have an unphysical value. This same type of issue arises when we interpolate/remap the state to a lat/lon grid for plotting.
- Solving the equations over the ocean: we need to set up initial conditions in regions where there is no land that wont break anything, and we evaluate tendencies over fake land in the ocean, which is inefficient, and these cant break anything
- Other: there are sometimes multiple masks (coming from different native resolutions of parameter files, e.g.). 

A related issue is that ultimately we would like to support lateral flow in the soil. This requires boundary conditions in the horizontal.

The benefits to what we are doing now is that (I think) regridding in the coupler is easy because land and atmos have the exact same space at the surface
## Priorities
- In the short term (i.e. done by this fall, for our paper), treating the land as a collection of independent columns is fine. We would not evaluate the model over ocean and not have to worry about setting values of parameters/IC over the ocean. That is, it would be nice to have a custom ClimaCore domain which is a set of independent columns which does not span the entire globe. Then we alleviate basically all of the issues above. 
- In the long term (end of next year?) supporting boundary conditions and lateral flow is a goal, but this is not essential and requires a lot more work. Note that there are two levels to this:

- - regional simulations (hybrid box - no mask required, but boundary conditions in the horizontal are required). Here we can learn from Parflow, e.g.,

- - global simulations (mask + horizontal BC) - can we learn from the ocean model/do what they do?

## Proposed Work Summer 2024
ClimaLand:
- short term: (what is in place now as of Aug 2024): set ocean values to mean over land, Ignore NaNs over the ocean that result as a function of the simulation. (on CPU, we might get breaking errors, while on GPU, running with NaNs does not break the sim)

ClimaCore:
- Define a new type of domain which is a set of independent columns not defined over the entire globe
- no DSS should be required and this should make the land model more efficient
-  Interpolation to the model grid needs to be mask aware
- How would the flux exchange need to adapt if the land model had a very different grid?

## Proposed Work 2025
- Enable boundary conditions for the horizontal using a mask. These BC would be applied where the mask transitions from 1->0 (land -> ocean). The prognostic variable fields would still be defined globally.

Notes from a while ago - need to revisit
When we have lateral flow we want to make use of the current topology/connectivity of points and how this is set up in ClimaCore. This is because it is easier to use a mask in boundary condition setting in the horizontal than to create a totally unstructured mesh.

I checked with Andre and he confirmed that this is what the ocean model does. there are prognostic variables over land but the tendencies are never evaluated. Only indices which are in the mask = 1 region are loaded onto the GPU.

**on the ClimaLand side:**
This solution means that we need to use the mask on the land side because we will still have fields defined over the ocean.

We will need to:
1. ingest the land/sea mask and store somewhere where it is accessible to the tendency 
2. initial state vector = 0 everywhere or we can set to NaN everywhere (?). 
3. Set initial conditions using the mask (NaN over ocean)
4. Parameter data should be masked already (NaN over the ocean). When regridding, we'll get back Nans in the ocean for the field. Interpolating to model grid needs to be mask aware
5. Write a macro which can be used in any broadcasted function when mask = 1, like:
```
#=
This assumes that `mask` is in local scope
=#
macro masked(expr)
    :(@. ifelse(mask == 0, 0, expr))
end

function tendency_func!(Y, p, t)
    (; mask) = p
    (; x, z) = p # fields
    (; ϕ, ψ) = Y # fields
   # @. ϕ = some_tendency(ψ, x, z) # current
  #  @. ϕ = ifelse(mask == 0, 0, some_tendency(ψ, x, z)) # what we want
    @masked ϕ = some_tendency(ψ, x, z) # cleaner way to implement
end
```
6. Repeat for all broadcasted expressions in the tendency, cache, etc.
7. use the mask in plotting (maybe not if NaN in ocean)

- is memory an issue since we are stored fields everywhere, when land is only 30% of the Earth's surface? We should be computed limited on GPU and not memory limited (this is true for the ocean model).

@sriharshakandala @juliasloan25 @Sbozzolo 
```[tasklist]
### Tasks
- [ ] Use ETOPO 2022 global relief model data to generate element-level `land-sea` mask
- [ ] Add `element mask` to ClimaCore spaces
- [ ] Block FD/SE/tendency operations for masked elements in ClimaCore
```

**Comment by kmdeck** [comment]
@charleskawczynski 
no immediate rush to answer, but this is something we'd like to figure out in the next month or two

**Comment by charleskawczynski** [comment]
My first thought was to bite the bullet and develop a special ClimaCore MaskedField, but the more that I think of it, the more I’m thinking that it’s easier to handle tendency case-by-case.
For example:

```julia
@. field_tendency = ifelse(mask==0, field_tendency, foo(Y.c))
```
We might be able to write a macro to write this more compactly, but I think this pretty effectively reuses our existing infrastructure. Also, this should still work with `foo(Y.c))` as a field or a function/expression. So, interpolations etc should still work. Using broadcast expressions is kind of important in terms of letting ClimaCore handle metric terms properly, so I think it’s important to not overlook this key feature.

I’m happy to chat more about this if I’ve missed something or if you think there may be more to the story.

**Comment by kmdeck** [comment]
Hi Charlie,

thanks for thinking about this! My concern with masking the tendency is that then we also have to mask the initial conditions or set IC over the ocean (ditto with all of our spatial parameter fields), and possibly use the mask in plotting (Im not sure the value of zero for parameters and the state will produce non-NaN tendencies, so we may not be able to set ocean-area values to zero and have it just work). it ends up getting kind of hacky.

we can discuss more! I can also show you our code and we can look at how complex it would be to implement the above^^

Our tendency functions have field vector arguments, so it didnt seem like they would broadcast as we want.

**Comment by charleskawczynski** [comment]
> we can discuss more! I can also show you our code and we can look at how complex it would be to implement the above^^

I think that would be really helpful

**Comment by kmdeck** [comment]
@simonbyrne, Tapio also suggested asking your opinion on this. He had the impression from you that restricting the grid to only include continents, while maintaining information about connectivity (to support tendencies in the horizontal, and, eventually, non-periodic boundary conditions), was not that difficult. 

If that is the case, it seems like the load-balancing problem would not arise, nor would the land model need to worry about about only evaluating tendencies over land, because our fields would not be defined over the ocean.

thank you!

**Comment by simone-silvestri** [comment]
Hi @kmdeck. 

In Oceananigans we have a structured mesh, so we maintain the structure but launch an unstructured kernel, i.e. we (1) map out the active cells in a linear map with `map[t] = (i, j, k)`, (2) launch a kernel with a number of threads equal to the total number of active cells and finally (3) map the unique thread index to the three-dimensional grid index inside the function.

In distributed computations, to load balance it is enough to calculate the total number of active cells in the global grid and divide them by processor. We do keep the memory associated with immersed cells though. In your case, it is probably better to go fully unstructured since you are using climacore that gives you this ability

**Comment by kmdeck** [comment]
@dennisYatunin, if we had NaNs as our field values over the ocean, and then tried to step Y.soil.variable implicitly, would we always use the max number of iterations over the ocean, since NaN < tolerance is always false?

**Comment by dennisYatunin** [comment]
Depends on how you have things set up. If you're using the convergence checker specified [here](https://github.com/CliMA/ClimaLand.jl/blob/9c97fded8c67c8d3f6d252c8e4a5aa61611bc6af/experiments/standalone/Soil/water_conservation.jl#L48-L49), then yes. You can also pass a custom `norm` to the convergence checker, though. The default is `LinearAlgebra.norm`, but you can define an alternative version that masks every field before reducing it with the standard L2-norm.

You can also just hardcode the number of Newton iterations, and avoid using a convergence checker altogether. That's what we do in ClimaAtmos.

**Comment by Sbozzolo** [comment]
It would be good if we could also skip the evaluation of external forcings (ie, read from file) on masked values too.

This is not as critical because it probably won't affect performance that much and it not causing issues NaNs.

**Comment by charleskawczynski** [comment]
The basic idea that I have in mind, in order, is:

 - Introduce a nodal mask into `SpectralElementGrid2D`, with a mechanism to allow users to update it somehow (e.g., `set_mask!(::Function, space::AbstractSpace)`).
 - Thread the mask through our broadcast functions, to avoid operations where the `mask` is 0:
   - `copyto!` (pointwise)
   - `apply_stencil!` (stencil operations)
   - more?

**Comment by Sbozzolo** [comment]
I agree it would be better to have a nodal mask. One of the challenges I see is with the diagnostics

> Leverage the mask to perform mask-aware horizontal interpolation. Here, we can use the mask values to determine how we want to perform interpolation / extrapolation.
How are you planning on tackling this?

The interpolation needed for the diagnostics is the spectral interpolation (in the Remapper), which is intrinsecally an element-wise operation. I genuinly don't know what's a good way to deal with this.

**Comment by charleskawczynski** [comment]
I'm not yet sure, but the most crud thing we could do that would be equivalent to the element-wise behavior would be to check the mask any one column within the element, and use that. At least with a nodal mask we have the possibility to do something more sophisticated.

This is not an uncommon problem, and I'm sure there's literature on this, we can look to see what other models do.

**Comment by charleskawczynski** [comment]
We need the mask to exist first, before we can do anything regarding horizontal interpolation. And simply supporting this for point-wise and stencil operations can make running the model easier.

**Comment by sriharshakandala** [comment]
In my opinion, adding an independent column space, either in `ClimaCore` or otherwise, would be the best option. I believe this is the most efficient and simple way do things as:

1. The current model seems to be 1D, which solves independent columns. We should not need to couple this with the current SE infrastructure, and then enforce masks, etc. It also ties this to the partitioning strategy of partitioning a cubed sphere which is also inefficient. This is unnecessarily complex. The alternative is fairly straightforward and quick to implement and does not contain any of the above restrictions and will be much more efficient and trivially parallelizable.

2). Even when we eventually add a 2D solver, they could do away with a simpler finite volume model or any other model of their choice. Their horizontal boundary conditions could be different (compared to no horizontal boundary conditions for our case). Their domain potentially consists of multiple disjoint domains and need different partitioning strategies for efficiency. It may not be best to squeeze everything into the current SE infrastructure.  We can also add these features in `ClimaCore`!

3). In both cases, we can use a function similar to SpaceVaryingInputs to transfer data between grids.

**Comment by charleskawczynski** [comment]
I hear you, @sriharshakandala, but this issue has been open for over a year, and we've discussed fixing this for a while. I would argue that the approach that I'm suggesting is much simpler, since it leverages our existing infrastructure. I spent yesterday working on it, and Kat and I are ready to try out https://github.com/CliMA/ClimaCore.jl/pull/2201 in a real example.

If your suggestion is 1 day of work, then please, by all means go ahead and add it to ClimaCore (we can always support both masks in our existing infrastructure + a new `IndependentColumns` type). But if the work that your suggesting is more than 1 day, which is what I suspect, then I think it would be wise to move forward with https://github.com/CliMA/ClimaCore.jl/pull/2201 and see how this goes.

**Comment by Sbozzolo** [comment]
> I'm not yet sure, but the most crud thing we could do that would be equivalent to the element-wise behavior would be to check the mask any one column within the element, and use that. At least with a nodal mask we have the possibility to do something more sophisticated.
> 
> This is not an uncommon problem, and I'm sure there's literature on this, we can look to see what other models do.

Okay, I'd be fine with starting with applying the mask over the entire element (through the columns) to sidestep this issue.

From a science and user fit point of view, I think @sriharshakandala 's solution would be preferable. Having to deal with spectral elements is a pointless complexity that makes everything harder. We wouldn't even need to talk about interpolation if we had columns on a regular lat-long grid. 

That said, in my opinion, the most important thing at this point is use well-suported and well-tested infrastructure, and keeping the infrastructure between ClimaLand and ClimaAtmos aligned is the easiest way to do so.

**Comment by charleskawczynski** [comment]
> From a science and user fit point of view, I think @sriharshakandala 's solution would be preferable.

How is that so? As I mentioned before, using an element mask will mean [that there will be land nodal stacks marked as ocean, and ocean nodal stacks marked as land](https://github.com/CliMA/ClimaLand.jl/issues/488#issuecomment-2683202853)...

**Comment by charleskawczynski** [comment]
This means that this mask solution will not fix one of the major existing annoyances of needing to define meaningless land parameters over the ocean.

**Comment by Sbozzolo** [comment]
I meant, the independent columns.

**Comment by charleskawczynski** [comment]
> I meant, the independent columns.

I agree that this would be ideal, assuming we would never need any horizontal operations.

**Comment by kmdeck** [comment]
The Independent columns approach is a good one, but it would be for a specific use case (land model with lateral flow off.this is a very important use case). Ultimately we would like to simulate lateral flow; this has been a goal since the beginning. It's up to the software team to decide if making something tailored/unique to the no lateral flow case is the right place to start, or if doing something like what Charlie has been working on is better because it can be used in both cases (no lateral flow/with lateral flow).

**Comment by sriharshakandala** [comment]
Sure. It may be fine with exploring this as a temporary solution, especially if this can be tried out in a couple of days! Elementwise mask, with placing each SE on a different CUDA block, might be easier!

**Comment by charleskawczynski** [comment]
Should we close this, @kmdeck?

**Comment by kmdeck** [comment]
woohoo! let's close it 🚀
